### PR TITLE
Add multiple stages for Azure Pipeline

### LINKED ISSUE #509: [FEATURE REQUEST]: Multiple stages for the Azure Pipeline
The build pipeline is reaching the timeout limit (one hour) after #506. We can break up the pipeline into multiple stages to avoid the timeout.

### PR DESCRIPTION
**Pull request by elvaliuliuliu**
This PR adds multiple stages for Azure Pipeline which will fix #509 , the Azure Pipeline will have the following stages:
* Build Sources: Build current sources
(Once the build is finished successfully, the following three testing stages will start)
* E2E Tests: Run current E2E tests
* E2E Forward Compatibility Tests: Run E2E tests in the lastest compatible release with current worker
* E2E Backward Compatibility Tests: Run current E2E tests with the lastest compatible worker

### CODE REVIEW

**Comment by imback82** [comment]
Can we not rebuild in the subsequent stages? We should just build once in the first stage and reuse the binaries in the testing stages.

**Comment by imback82** [comment]
Is this ready for review?

**Comment by elvaliuliuliu** [comment]
> Is this ready for review?

Yea, this is ready for review. Thanks!

File: azure-pipelines.yml [hunk 1]
```diff
@@ -11,9 +11,9 @@ variables:
   _SignType: real
   _TeamName: DotNetSpark
   MSBUILDSINGLELOADCONTEXT: 1
-  # backwardCompatibleRelease/forwardCompatibleRelease is the "oldest" releases that work with the current release
-  backwardCompatibleRelease: '0.9.0'
+  # forwardCompatibleRelease/backwardCompatibleRelease is the "oldest" releases that work with the current release
   forwardCompatibleRelease: '0.9.0'
+  backwardCompatibleRelease: '0.9.0'
```
Comment (imback82) [inline azure-pipelines.yml:16]: any reason the order got flipped?

File: azure-pipelines.yml [hunk 2]
```diff
@@ -22,7 +22,7 @@ variables:
   (FullyQualifiedName!=Microsoft.Spark.E2ETest.UdfTests.UdfSimpleTypesTests.TestUdfWithReturnAsTimestampType)&\
   (FullyQualifiedName!=Microsoft.Spark.E2ETest.UdfTests.UdfSimpleTypesTests.TestUdfWithTimestampType)&\
   (FullyQualifiedName!=Microsoft.Spark.E2ETest.IpcTests.SparkSessionTests.TestCreateDataFrameWithTimestamp)"
-  LatestDotnetWorkerDir: '$(Build.ArtifactStagingDirectory)\Microsoft.Spark.Worker\netcoreapp3.1\win-x64'
+  CurrentDotnetWorkerDir: '$(Build.ArtifactStagingDirectory)\Microsoft.Spark.Binaries\Microsoft.Spark.Worker\netcoreapp3.1\win-x64'
```
Comment (imback82) [inline azure-pipelines.yml:25]: I see "$(Build.ArtifactStagingDirectory)/Microsoft.Spark.Binaries" in many places. Can you create a variable for this?

File: azure-pipelines.yml [hunk 3]
```diff
@@ -38,436 +38,505 @@ resources:
...
+      inputs:
+        sourceFolder: $(Build.ArtifactStagingDirectory)/Microsoft.Spark.Binaries/Jars
+        contents: '**/microsoft-spark-2.3.x**'
+        targetFolder: $(Build.SourcesDirectory)/src/scala
+
+    - task: CopyFiles@2
+      displayName: Copy microsoft-spark-2.4.x.jar
+      inputs:
+        sourceFolder: $(Build.ArtifactStagingDirectory)/Microsoft.Spark.Binaries/Jars
+        contents: '**/microsoft-spark-2.4.x**'
```
Comment (imback82) [inline azure-pipelines.yml:185]: is it possible to have `**/microsoft-spark-**` or `**/*.jar` or something similar to copy both versions in one task?

File: azure-pipelines.yml [hunk 4]
```diff
@@ -38,436 +38,505 @@ resources:
...
+                     /p:SparkWorkerPublishDir=$(Build.ArtifactStagingDirectory)\Microsoft.Spark.Binaries\Microsoft.Spark.Worker
+                     /p:SparkWorkerPackageOutputDir=$(Build.ArtifactStagingDirectory)\Microsoft.Spark.Binaries
+
+      - task: PublishBuildArtifacts@1
+        inputs:
+          pathtoPublish: '$(Build.ArtifactStagingDirectory)/Microsoft.Spark.Binaries'
+          artifactName:  Microsoft.Spark.Binaries
+
+- stage: Test
+  displayName: Test Sources
```
Comment (imback82) [inline azure-pipelines.yml:155]: How about we use the followings for each stage's display name:
- E2E Tests
- E2E Forward Compatibility Tests
- E2E Backward Compatibility Tests

And each job name under stage can be just `Run`

**Comment by imback82** [comment]
Btw, the azure-pipeline.xml is getting too long to maintain. Can you update this with templates, https://docs.microsoft.com/en-us/azure/devops/pipelines/process/templates?view=azure-devops? E2E tests are good candidates. You can do this as a follow-up PR.

File: azure-pipelines.yml (hunk 1 above)
Comment (elvaliuliuliu) [reply azure-pipelines.yml:16]: I just thought flip the order make it more readable. But I can get it back. Please let me know what is preferred.

File: azure-pipelines.yml (hunk 2 above)
Comment (elvaliuliuliu) [reply azure-pipelines.yml:25]: Do you mean to create two variables one with `/`, another with `\`?

File: azure-pipelines.yml (hunk 3 above)
Comment (elvaliuliuliu) [reply azure-pipelines.yml]: Sure, changed.

File: azure-pipelines.yml (hunk 4 above)
Comment (elvaliuliuliu) [reply azure-pipelines.yml]: Sounds good, changed the name.

**Comment by elvaliuliuliu** [comment]
> Btw, the azure-pipeline.xml is getting too long to maintain. Can you update this with templates, https://docs.microsoft.com/en-us/azure/devops/pipelines/process/templates?view=azure-devops? E2E tests are good candidates. You can do this as a follow-up PR.

Sounds good, E2E tests could be a template. Will do as a follow-up PR.

File: azure-pipelines.yml (hunk 2 above)
Comment (imback82) [reply azure-pipelines.yml:25]: Can't they be used interchangeably?

File: azure-pipelines.yml (hunk 1 above)
Comment (imback82) [reply azure-pipelines.yml:16]: ok, not sure why it's more readable, but it's fine.

File: azure-pipelines.yml (hunk 2 above)
Comment (elvaliuliuliu) [reply azure-pipelines.yml]: Added.

**Pull request merged**